### DISCUSSION TITLE
Tips for debugging abort with multithreading

### DISCUSSION THREAD

**Discussion by rofinn** [duckdb/duckdb#4843] [Q&A]
Hi there,

I'm currently getting the following SIGABRT error from duckdb. The Julia code itself isn't using any threading, but the error only occurs if `threads > 1`. Before I start dumping SQL queries and running them through `gdb` do you have any recommendations of things I should try first? Maybe there's a way for me to produce an error message with more info about the query result state before it just exits?

```
libc++abi: terminating with uncaught exception of type duckdb::InvalidInputException: Invalid Input Error: Attempting to execute an unsuccessful or closed pending query result

signal (6): Abort trap: 6
in expression starting at REPL[40]:1
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
pthread_kill at /usr/lib/system/libsystem_pthread.dylib (unknown line)
abort at /usr/lib/system/libsystem_c.dylib (unknown line)
abort_message at /usr/lib/libc++abi.dylib (unknown line)
_ZL28demangling_terminate_handlerv at /usr/lib/libc++abi.dylib (unknown line)
_ZL15_objc_terminatev at /usr/lib/libobjc.A.dylib (unknown line)
_ZSt11__terminatePFvvE at /usr/lib/libc++abi.dylib (unknown line)
_ZN10__cxxabiv1L12failed_throwEPNS_15__cxa_exceptionE at /usr/lib/libc++abi.dylib (unknown line)
__cxa_throw at /usr/lib/libc++abi.dylib (unknown line)
_ZN6duckdb18PendingQueryResult23CheckExecutableInternalERNS_17ClientContextLockE at /Users/rory/.julia/artifacts/55d0c979d485f2b1eb30e6f13ff4b5ebe8834ddc/lib/libduckdb.dylib (unknown line)
_ZN6duckdb18PendingQueryResult15ExecuteInternalERNS_17ClientContextLockE at /Users/rory/.julia/artifacts/55d0c979d485f2b1eb30e6f13ff4b5ebe8834ddc/lib/libduckdb.dylib (unknown line)
_ZN6duckdb18PendingQueryResult7ExecuteEv at /Users/rory/.julia/artifacts/55d0c979d485f2b1eb30e6f13ff4b5ebe8834ddc/lib/libduckdb.dylib (unknown line)
duckdb_execute_pending at /Users/rory/.julia/artifacts/55d0c979d485f2b1eb30e6f13ff4b5ebe8834ddc/lib/libduckdb.dylib (unknown line)
duckdb_execute_pending at /Users/rory/.julia/packages/DuckDB/2OSsm/src/api.jl:1290 [inlined]
execute at /Users/rory/.julia/packages/DuckDB/2OSsm/src/result.jl:613
execute at /Users/rory/.julia/packages/DuckDB/2OSsm/src/result.jl:696 [inlined]
execute at /Users/rory/.julia/packages/DBInterface/1Gmxx/src/DBInterface.jl:130 [inlined]
#execute#2 at /Users/rory/.julia/packages/DBInterface/1Gmxx/src/DBInterface.jl:152 [inlined]
execute at /Users/rory/.julia/packages/DBInterface/1Gmxx/src/DBInterface.jl:152 [inlined]
```

**Comment by hannes** [comment]
Can you try to make a reproducible example please and file an issue?

**Comment by Cauchy-NY** [comment]
@rofinn Hi, I have met the same problem recently, and after searching this hint in the codebase of DuckDB, I found this functions.

```
unique_ptr<ClientContextLock> StreamQueryResult::LockContext() {
	if (!context) {
		string error_str = "Attempting to execute an unsuccessful or closed pending query result";
		if (HasError()) {
			error_str += StringUtil::Format("\nError: %s", GetError());
		}
		throw InvalidInputException(error_str);
	}
	return context->LockContext();
}

void StreamQueryResult::CheckExecutableInternal(ClientContextLock &lock) {
	if (!IsOpenInternal(lock)) {
		string error_str = "Attempting to execute an unsuccessful or closed pending query result";
		if (HasError()) {
			error_str += StringUtil::Format("\nError: %s", GetError());
		}
		throw InvalidInputException(error_str);
	}
}
```


You should avoid sharing connections between threads, since the activity on the connection will mean that only one thread will be able to do anything at a time.

The cost of creating a TCP connection is expensive. Meanwhile, the number of concurrent TCP connections that the DB server can carry is also limited. Therefore, it is unrealistic to create a Connection every time it is called. so we need a database connection pool.

**Comment by Cauchy-NY** [comment]
@Mause Hello Mause, I have two questions about DuckDB
1. In DuckDB docs, "For DuckDB, there is no DBMS server software to install, update and maintain", so does a connection in DuckDB still correspond to a TCP link?
2. I know DuckDB provides a duplicate() function now to create multiple read-write connections to the same database file or the same in-memory database instance. Do you plan to support database connection pool in the future?

**Reply by Mause** [reply]
There are no TCP connections, this is an embedded database, hence no server or client

**Comment by rofinn** [comment]
Hmm, yeah, there shouldn't be any TCP connections, but I wonder if there's still and issue with sharing the same connection object across threads... maybe I need to explicitly define a `cursor` object for each thread?

**Reply by Mytherin** [reply]
Ideally a separate connection object is used for each thread, indeed. Sharing a connection object should work, but there is a mutex involved that prevents any sort of speed-ups.
